nixos/modules/services/misc/bees.nix

```diff
@@ -0,0 +1,94 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  inherit (lib) mkOption types optionalString;
+  cfg = config.services.beesd;
+
+  fsOptions = {...}:
+  with types; {
+    options.spec = mkOption {
+      type = string;
+      description = ''
+        Specification usable by findmnt; ie. LABEL=foo, UUID=bar, etc.
+      '';
+    };
+    options.hashTableSizeMB = mkOption {
+      type = enum [ 16 32 64 128 256 512 1024 2048 4096 8192 16384 32768 65536 131072 262144 524288 1048576 2097152 4194304 8388608 ];
+      default = 1024; # 1GB; default from upstream beesd script
+      description = ''
+        Hash table size; must be 16MB*(2^n).
+
+        A larger ratio of index size to storage size means smaller blocks of duplicate content are recognized.
+
+        If you have 1TB of data, a 4GB hash table (which is to say, a value of 4096 -- 16MB*(2^8)) will permit 4KB extents (the smallest possible size) to be recognized, whereas a value of 1024 -- creating a 1GB hash table -- will recognize only aligned duplicate blocks of 16KB.
+      '';
+    };
+    options.verbosity = mkOption {
+      type = int;
+      default = 6;
+      description = "Log verbosity (0-8)";
+    };
+    options.workDir = mkOption {
+      type = str;
+      default = ".beeshome";
+      description = ''
+        Name (relative to the root of the filesystem) of the subvolume where the hash table will be stored.
+      '';
+    };
+  };
+
+in {
+
+  options.services.beesd = {
+    filesystems = mkOption {
+      description = "BTRFS filesystems to run block-level deduplication on";
+      default = { };
+      example = literalExample {
+        root = {
+          spec = "LABEL=root";
+          hashTableSizeMB = 2048;
+          verbosity = 2;
+        };
+      };
+      type = with types; attrsOf (submodule fsOptions);
+    };
+  };
+  config = {
+    systemd.services = mapAttrs' (name: fs: nameValuePair "beesd@${name}" {
+      description = "Block-level BTRFS deduplication for %i";
+      after = [ "sysinit.target" ];
+
+      serviceConfig = {
+        ExecStart="${pkgs.bees}/bin/bees-service-wrapper run %i";
+        ExecStopPost="${pkgs.bees}/bin/bees-service-wrapper cleanup %i";
+        Nice = 19;
+        KillMode = "control-group";
+        KillSignal = "SIGTERM";
+        CPUShares = 128;
+        StartupCPUShares = 256;
+        BlockIOWeight = 100;
+        StartupBlockIOWeight = 250;
+        IOSchedulingClass = "idle";
+        IOSchedulingPriority = 7;
```

Comment:
Where do these values come from? If it comes from an upstream file, you could link it here.